File: frontend/lib/features/games/data/models/game_model.dart
import 'package:intl/intl.dart';

/// Full game/match model mirroring the backend Game schema.
class GameModel {
  const GameModel({
    required this.id,
    required this.title,
    required this.sport,
    required this.organizer,
    required this.scheduledAt,
    required this.durationMinutes,
    required this.location,
    required this.maxPlayers,
    required this.players,
    required this.requiredSkillLevel,
    required this.status,
    this.description,
    this.groupChatId,
    this.tags = const [],
    this.result,
    this.createdAt,
  });

  final String id;
  final String title;
  final String? description;
  final String sport;
  final GameOrganizer organizer;
  final DateTime scheduledAt;
  final int durationMinutes;
  final GameLocation location;
  final int maxPlayers;
  final List<GamePlayer> players;
  final String requiredSkillLevel;
  final String status;
  final String? groupChatId;
  final List<String> tags;
  final GameResult? result;
  final DateTime? createdAt;

  // ── Computed ──────────────────────────────────────────────
  int get approvedCount =>
      players.where((p) => p.status == 'approved').length;
  int get pendingCount =>
      players.where((p) => p.status == 'pending').length;
  int get availableSlots => maxPlayers - approvedCount;
  bool get isFull => availableSlots <= 0;
  bool get isOpen => status == 'open';
  bool get isUpcoming => status == 'open' || status == 'full';
  bool get isInProgress => status == 'in_progress';
  bool get isCompleted => status == 'completed';
  bool get isCancelled => status == 'cancelled';
  DateTime get endTime =>
      scheduledAt.add(Duration(minutes: durationMinutes));

  String get formattedDate =>
      DateFormat('EEE, d MMM y').format(scheduledAt);
  String get formattedTime => DateFormat('HH:mm').format(scheduledAt);
  String get formattedDuration {
    final h = durationMinutes ~/ 60;
    final m = durationMinutes % 60;
    if (h == 0) return '${m}min';
    if (m == 0) return '${h}h';
    return '${h}h ${m}min';
  }

  bool hasPlayer(String userId) =>
      players.any((p) => p.userId == userId);

  GamePlayer? getPlayer(String userId) {
    final matches = players.where((p) => p.userId == userId).toList();
    return matches.isEmpty ? null : matches.first;
  }

  // ── Serialisation ─────────────────────────────────────────
  static String? _parseGroupChatId(dynamic value) {
    if (value == null) return null;
    if (value is String) return value;
    if (value is Map) {
      final id = value['_id'] ?? value['id'];
      return id is String ? id : id?.toString();
    }
    return null;
  }

  factory GameModel.fromJson(Map<String, dynamic> json) {
    return GameModel(
      id: json['_id'] as String? ?? json['id'] as String,
      title: json['title'] as String,
      description: json['description'] as String?,
      sport: json['sport'] as String,
      organizer: GameOrganizer.fromJson(
          json['organizer'] as Map<String, dynamic>),
      scheduledAt:
          DateTime.parse(json['scheduledAt'] as String),
      durationMinutes: json['durationMinutes'] as int? ?? 60,
      location: GameLocation.fromJson(
          json['location'] as Map<String, dynamic>),
      maxPlayers: json['maxPlayers'] as int,
      players: (json['players'] as List<dynamic>?)
              ?.map((e) =>
                  GamePlayer.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      requiredSkillLevel:
          json['requiredSkillLevel'] as String? ?? 'any',
      status: json['status'] as String? ?? 'open',
      groupChatId: _parseGroupChatId(json['groupChat']),
      tags: (json['tags'] as List<dynamic>?)
              ?.cast<String>() ??
          [],
      result: json['result'] != null
          ? GameResult.fromJson(
              json['result'] as Map<String, dynamic>)
          : null,
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'title': title,
        'sport': sport,
        if (description != null) 'description': description,
        'scheduledAt': scheduledAt.toIso8601String(),
        'durationMinutes': durationMinutes,
        'location': location.toJson(),
        'maxPlayers': maxPlayers,
        'requiredSkillLevel': requiredSkillLevel,
        if (tags.isNotEmpty) 'tags': tags,
      };

  GameModel copyWith({
    String? title,
    String? description,
    String? sport,
    DateTime? scheduledAt,
    int? durationMinutes,
    GameLocation? location,
    int? maxPlayers,
    List<GamePlayer>? players,
    String? requiredSkillLevel,
    String? status,
  }) {
    return GameModel(
      id: id,
      title: title ?? this.title,
      description: description ?? this.description,
      sport: sport ?? this.sport,
      organizer: organizer,
      scheduledAt: scheduledAt ?? this.scheduledAt,
      durationMinutes: durationMinutes ?? this.durationMinutes,
      location: location ?? this.location,
      maxPlayers: maxPlayers ?? this.maxPlayers,
      players: players ?? this.players,
      requiredSkillLevel:
          requiredSkillLevel ?? this.requiredSkillLevel,
      status: status ?? this.status,
      groupChatId: groupChatId,
      tags: tags,
      result: result,
      createdAt: createdAt,
    );
  }

  @override
  bool operator ==(Object other) =>
      other is GameModel && other.id == id;

  @override
  int get hashCode => id.hashCode;
}

// ── Sub-models ────────────────────────────────────────────────────────────────
class GameOrganizer {
  const GameOrganizer({
    required this.id,
    required this.username,
    required this.firstName,
    required this.lastName,
    this.profilePicture,
  });

  final String id;
  final String username;
  final String firstName;
  final String lastName;
  final String? profilePicture;

  String get fullName => '$firstName $lastName';

  factory GameOrganizer.fromJson(Map<String, dynamic> json) =>
      GameOrganizer(
        id: json['_id'] as String? ?? json['id'] as String,
        username: json['username'] as String,
        firstName: json['firstName'] as String,
        lastName: json['lastName'] as String,
        profilePicture: json['profilePicture'] as String?,
      );
}

class GameLocation {
  const GameLocation({
    required this.city,
    this.neighborhood,
    this.venueName,
  });

  final String city;
  final String? neighborhood;
  final String? venueName;

  String get displayName {
    final parts = <String>[];
    if (venueName != null) parts.add(venueName!);
    if (neighborhood != null) parts.add(neighborhood!);
    parts.add(city);
    return parts.join(', ');
  }

  factory GameLocation.fromJson(Map<String, dynamic> json) =>
      GameLocation(
        city: json['city'] as String,
        neighborhood: json['neighborhood'] as String?,
        venueName: json['venueName'] as String?,
      );

  Map<String, dynamic> toJson() => {
        'city': city,
        if (neighborhood != null) 'neighborhood': neighborhood,
        if (venueName != null) 'venueName': venueName,
      };
}

class GamePlayer {
  const GamePlayer({
    required this.userId,
    required this.username,
    required this.status,
    this.firstName,
    this.lastName,
    this.profilePicture,
    this.joinedAt,
  });

  final String userId;
  final String username;
  final String status; // approved | pending | invited | kicked | left
  final String? firstName;
  final String? lastName;
  final String? profilePicture;
  final DateTime? joinedAt;

  bool get isApproved => status == 'approved';
  bool get isPending => status == 'pending';

  String get displayName =>
      (firstName != null && lastName != null)
          ? '$firstName $lastName'
          : username;

  factory GamePlayer.fromJson(Map<String, dynamic> json) {
    final userRaw = json['user'];
    final Map<String, dynamic>? user = userRaw is Map<String, dynamic>
        ? userRaw
        : null;
    // Backend may send user as String (e.g. id reference) or as populated map
    final String userId = user != null
        ? (user['_id'] as String? ?? user['id'] as String? ?? '')
        : (userRaw is String
            ? userRaw
            : (json['userId'] as String? ?? ''));
    final String username =
        user?['username'] as String? ?? json['username'] as String? ?? '';

    return GamePlayer(
      userId: userId,
      username: username,
      status: json['status'] as String,
      firstName: user?['firstName'] as String?,
      lastName: user?['lastName'] as String?,
      profilePicture: user?['profilePicture'] as String?,
      joinedAt: json['joinedAt'] != null
          ? DateTime.tryParse(json['joinedAt'] as String)
          : null,
    );
  }
}

class GameResult {
  const GameResult({
    this.winner,
    this.score,
    this.notes,
    this.completedAt,
  });

  final String? winner;
  final String? score;
  final String? notes;
  final DateTime? completedAt;

  factory GameResult.fromJson(Map<String, dynamic> json) =>
      GameResult(
        winner: json['winner'] as String?,
        score: json['score'] as String?,
        notes: json['notes'] as String?,
        completedAt: json['completedAt'] != null
            ? DateTime.tryParse(json['completedAt'] as String)
            : null,
      );
}

// ── Search parameters ─────────────────────────────────────────────────────────
class GameSearchParams {
  const GameSearchParams({
    this.sport,
    this.city,
    this.skillLevel,
    this.status = 'open',
    this.page = 1,
    this.limit = 20,
    this.sortBy = 'scheduledAt',
  });

  final String? sport;
  final String? city;
  final String? skillLevel;
  final String status;
  final int page;
  final int limit;
  final String sortBy;

  Map<String, dynamic> toQueryParams() => {
        if (sport != null) 'sport': sport,
        if (city != null) 'city': city,
        if (skillLevel != null) 'skillLevel': skillLevel,
        'status': status,
        'page': page,
        'limit': limit,
        'sortBy': sortBy,
      };

  GameSearchParams copyWith({
    String? sport,
    String? city,
    String? skillLevel,
    String? status,
    int? page,
  }) {
    return GameSearchParams(
      sport: sport ?? this.sport,
      city: city ?? this.city,
      skillLevel: skillLevel ?? this.skillLevel,
      status: status ?? this.status,
      page: page ?? this.page,
      limit: limit,
      sortBy: sortBy,
    );
  }

  GameSearchParams clearSport() => GameSearchParams(
        city: city,
        skillLevel: skillLevel,
        status: status,
      );

  @override
  bool operator ==(Object other) =>
      other is GameSearchParams &&
      other.sport == sport &&
      other.city == city &&
      other.skillLevel == skillLevel &&
      other.status == status &&
      other.page == page;

  @override
  int get hashCode =>
      Object.hash(sport, city, skillLevel, status, page);
}

File: backend/src/models/Game.model.js
const mongoose = require('mongoose');

// ─────────────────────────────────────────────
//  Sub-schemas
// ─────────────────────────────────────────────

/**
 * Represents a single player slot within a game.
 * Status transitions: invited → pending → approved | kicked | left
 */
const playerSlotSchema = new mongoose.Schema(
  {
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    status: {
      type: String,
      enum: ['invited', 'pending', 'approved', 'kicked', 'left'],
      default: 'pending',
    },
    role: {
      type: String,
      enum: ['player', 'co-organizer'],
      default: 'player',
    },
    joinedAt: { type: Date, default: Date.now },
    // Timestamp for when the organizer approved/kicked the player
    resolvedAt: { type: Date },
  },
  { _id: false },
);

/**
 * Area-level location — never precise GPS coordinates.
 * postalCode is selected: false to prevent accidental exposure.
 */
const gamLocationSchema = new mongoose.Schema(
  {
    venueName: { type: String, trim: true, maxlength: 100 },
    address: { type: String, trim: true, maxlength: 200 },
    neighborhood: {
      type: String,
      trim: true,
      required: [true, 'Neighborhood is required for matchmaking'],
    },
    city: {
      type: String,
      trim: true,
      required: [true, 'City is required'],
    },
    state: { type: String, trim: true },
    country: { type: String, trim: true, default: 'US' },
    postalCode: { type: String, trim: true, select: false },
  },
  { _id: false },
);

const gameResultSchema = new mongoose.Schema(
  {
    winnerDescription: { type: String, trim: true },
    score: { type: String, trim: true },
    mvpUser: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
    notes: { type: String, trim: true, maxlength: 500 },
    recordedAt: { type: Date, default: Date.now },
    recordedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
  },
  { _id: false },
);

// ─────────────────────────────────────────────
//  Main Game Schema
// ─────────────────────────────────────────────

const gameSchema = new mongoose.Schema(
  {
    // ── Core info ─────────────────────────────────────────────
    title: {
      type: String,
      required: [true, 'Game title is required'],
      trim: true,
      maxlength: [100, 'Title cannot exceed 100 characters'],
    },
    description: {
      type: String,
      trim: true,
      maxlength: [1000, 'Description cannot exceed 1000 characters'],
    },
    sport: {
      type: String,
      required: [true, 'Sport type is required'],
      trim: true,
      lowercase: true,
    },
    tags: [{ type: String, trim: true, lowercase: true }],

    // ── Organizer ─────────────────────────────────────────────
    organizer: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: [true, 'Organizer is required'],
      index: true,
    },

    // ── Schedule ──────────────────────────────────────────────
    scheduledAt: {
      type: Date,
      required: [true, 'Scheduled date/time is required'],
      validate: {
        validator(date) {
          return date > new Date();
        },
        message: 'Game must be scheduled in the future',
      },
    },
    /**
     * Duration in minutes.
     * Used for double-booking prevention:
     *   occupiedUntil = scheduledAt + durationMinutes
     */
    durationMinutes: {
      type: Number,
      required: [true, 'Duration is required'],
      min: [15, 'Game must last at least 15 minutes'],
      max: [480, 'Game cannot exceed 8 hours'],
    },

    // ── Location ──────────────────────────────────────────────
    location: {
      type: gamLocationSchema,
      required: true,
    },

    // ── Capacity ──────────────────────────────────────────────
    maxPlayers: {
      type: Number,
      required: [true, 'Maximum players is required'],
      min: [2, 'A game must allow at least 2 players'],
      max: [100, 'A game cannot exceed 100 players'],
    },
    minPlayersToStart: {
      type: Number,
      default: 2,
      min: 2,
    },

    // ── Players ───────────────────────────────────────────────
    players: {
      type: [playerSlotSchema],
      default: [],
    },

    // ── Skill filter ──────────────────────────────────────────
    requiredSkillLevel: {
      type: String,
      enum: ['any', 'beginner', 'intermediate', 'advanced', 'professional'],
      default: 'any',
    },

    // ── Game status ───────────────────────────────────────────
    /**
     * Lifecycle:  draft → open ─┬─ full ──┬─ in_progress → completed
     *                            │         └─ (capacity freed) → open
     *                            └─ cancelled
     */
    status: {
      type: String,
      enum: {
        values: ['draft', 'open', 'full', 'in_progress', 'completed', 'cancelled'],
        message: 'Invalid game status',
      },
      default: 'open',
      index: true,
    },
    cancelledReason: { type: String, trim: true },
    cancelledAt: { type: Date },
    cancelledBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },

    // ── Visibility & join settings ────────────────────────────
    isPrivate: { type: Boolean, default: false },
    requiresApproval: { type: Boolean, default: false },
    allowSpectators: { type: Boolean, default: true },

    // ── Group merge support ───────────────────────────────────
    /**
     * When two under-capacity games are merged:
     *  - The surviving game sets mergedWith = [id of absorbed game]
     *  - The absorbed game sets mergedInto = id of surviving game
     *    and status = 'cancelled'
     */
    isMerged: { type: Boolean, default: false },
    mergedWith: [{ type: mongoose.Schema.Types.ObjectId, ref: 'Game' }],
    mergedInto: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Game',
      default: null,
    },
    mergedAt: { type: Date },

    // ── Chat room ─────────────────────────────────────────────
    groupChat: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Chat',
      default: null,
    },

    // ── Post-game result ──────────────────────────────────────
    result: { type: gameResultSchema, default: null },

    // ── Soft delete ───────────────────────────────────────────
    deletedAt: { type: Date, default: null },
  },
  {
    timestamps: true,
    toJSON: { virtuals: true },
    toObject: { virtuals: true },
  },
);

// ─────────────────────────────────────────────
//  Virtuals
// ─────────────────────────────────────────────

gameSchema.virtual('approvedPlayerCount').get(function () {
  return this.players.filter((p) => p.status === 'approved').length;
});

gameSchema.virtual('availableSlots').get(function () {
  const taken = this.players.filter((p) => p.status === 'approved').length;
  return Math.max(0, this.maxPlayers - taken);
});

/**
 * Returns the Date when the game is expected to end.
 * Used for double-booking validation.
 */
gameSchema.virtual('occupiedUntil').get(function () {
  if (!this.scheduledAt || !this.durationMinutes) return null;
  return new Date(this.scheduledAt.getTime() + this.durationMinutes * 60 * 1000);
});

gameSchema.virtual('isUpcoming').get(function () {
  return this.scheduledAt > new Date() && ['open', 'full', 'draft'].includes(this.status);
});

// ─────────────────────────────────────────────
//  Indexes
// ─────────────────────────────────────────────

// Core matchmaking query: sport + city + status + date
gameSchema.index({ sport: 1, 'location.city': 1, status: 1, scheduledAt: 1 });

// Neighbourhood-level search
gameSchema.index({ 'location.neighborhood': 1, sport: 1, scheduledAt: 1 });

// Organizer's game list
gameSchema.index({ organizer: 1, status: 1 });

// Skill level filtering
gameSchema.index({ requiredSkillLevel: 1, status: 1 });

// Calendar queries (date range)
gameSchema.index({ scheduledAt: 1, status: 1 });

// Soft-delete filter
gameSchema.index({ deletedAt: 1 });

// Text search on title and description
gameSchema.index({ title: 'text', description: 'text', sport: 'text', tags: 'text' });

// ─────────────────────────────────────────────
//  Pre-save Hooks
// ─────────────────────────────────────────────

/**
 * Auto-update status to 'full' when approved player count hits maxPlayers,
 * and revert to 'open' if a player leaves and a slot opens up.
 */
gameSchema.pre('save', function (next) {
  if (!this.isModified('players') && !this.isModified('maxPlayers')) return next();

  const approvedCount = this.players.filter((p) => p.status === 'approved').length;

  if (this.status === 'open' && approvedCount >= this.maxPlayers) {
    this.status = 'full';
  } else if (this.status === 'full' && approvedCount < this.maxPlayers) {
    this.status = 'open';
  }

  return next();
});

// ─────────────────────────────────────────────
//  Instance Methods
// ─────────────────────────────────────────────

/**
 * Checks whether a user already has an approved slot in this game.
 */
gameSchema.methods.hasPlayer = function (userId) {
  return this.players.some((p) => p.user.toString() === userId.toString() && p.status === 'approved');
};

/**
 * Detects a schedule conflict for a given user across all their games.
 * Call this as a static before adding a player to a new game.
 */
gameSchema.statics.hasConflict = async function (userId, proposedStart, proposedDurationMinutes) {
  const proposedEnd = new Date(proposedStart.getTime() + proposedDurationMinutes * 60 * 1000);

  const conflicting = await this.findOne({
    'players.user': userId,
    'players.status': 'approved',
    status: { $in: ['open', 'full', 'in_progress'] },
    deletedAt: null,
    // Overlap condition: existing.start < proposed.end AND existing.end > proposed.start
    scheduledAt: { $lt: proposedEnd },
    $expr: {
      $gt: [
        { $add: ['$scheduledAt', { $multiply: ['$durationMinutes', 60000] }] },
        proposedStart.getTime(),
      ],
    },
  });

  return !!conflicting;
};

/**
 * Returns a lean summary object safe for public API responses.
 * Excludes sensitive or internal fields.
 */
gameSchema.methods.toPublicJSON = function () {
  const obj = this.toObject();
  delete obj.deletedAt;
  delete obj.__v;
  return obj;
};

// ─────────────────────────────────────────────
//  Query Helpers
// ─────────────────────────────────────────────

gameSchema.query.upcoming = function () {
  return this.where({ scheduledAt: { $gt: new Date() }, deletedAt: null });
};

gameSchema.query.active = function () {
  return this.where({ status: { $in: ['open', 'full', 'in_progress'] }, deletedAt: null });
};

// ─────────────────────────────────────────────
//  Model Export
// ─────────────────────────────────────────────

const Game = mongoose.model('Game', gameSchema);

module.exports = Game;
